### PR TITLE
[Bugfix] Autotune FlashInfer deferred MoE decode kernels before CUDA graph capture

### PR DESCRIPTION
**Pull request by mingg26**
## Purpose

Split out the MoE half of #54507 into a standalone fix. This is not duplicate work: #54507 covers both the MLA and MoE decode-autotune cache-miss bugs, while this PR isolates only the MoE part so it can be reviewed and merged independently.

Kimi-K3 decode defers MoE finalization so the consumer can fuse the final top-k reduction. That changes the monolithic FlashInfer MoE output signature from `(tokens, hidden)` to `(tokens, 0)`. The existing generic full-model autotune dummy does not exercise this deferred-finalize path, so FULL decode graph capture misses the `(B, 0)` cache entries, falls back to `MoERunner tactic=-1`, and captures that default tactic into the serving graph.

This change collects `defer_moe_finalize_max_num_tokens` from live deferred-finalize MoE layers and appends the bounded values to the existing FlashInfer autotune token counts. Each token count runs through the normal full-model dummy path with `get_hybrid_num_tokens_buckets(num_tokens)`, so the deferred path is exercised without any MoE-specific dispatcher calls.

The maximum-token pass runs first and tunes the widest bucket set. The later bounded passes reuse those cache entries for unchanged FlashInfer operations because the cache key contains the mapped profile, not the complete bucket list. Only signatures reached through the small-token control flow, including the deferred MoE `(B, 0)` signature, require additional tuning. The remaining overhead is the ordinary smaller full-model forward and cache lookup.

In the K3 B12 validation in #54507, the tuned path and the `tactic=-1` fallback launch exactly the same two MoE BMM kernels under Nsight Systems. This fix therefore changes the autotune result from a miss to a hit without changing the selected MoE kernels in that workload; it is performance-neutral there while removing a real cache-miss/default-tactic bug.

## Test Plan

1. Unit: `.venv/bin/python -m pytest tests/model_executor/test_flashinfer_autotune_warmup.py -q`
2. Static checks: `pre-commit run --files vllm/model_executor/warmup/kernel_warmup.py vllm/utils/flashinfer.py tests/model_executor/test_flashinfer_autotune_warmup.py`
3. Hardware cache-hit check: serve a truncated Kimi-K3 (4 layers, full 896 experts, dummy weights) on 2x4 GB300, TP8, `cudagraph_mode=FULL`, `--enable-flashinfer-autotune`, and `FLASHINFER_LOGGING_LEVEL=DEBUG`; count `[AutoTuner]: Using fallback tactic` lines for `trtllm_fp4_block_scale_moe` during final graph capture before and after this fix.

## Test Result

Unit:

```text
test_flashinfer_autotune_token_counts_include_deferred_moe_limits PASSED
test_flashinfer_autotune_token_counts_are_bounded_and_deduplicated PASSED
test_flashinfer_autotune_uses_token_buckets_for_each_dummy_run PASSED
3 passed
```

Hardware cache-hit validation from the initial implementation (fallback-tactic lines during final FULL graph capture, rank 0):

| op | before | after |
|---|---:|---:|
| `trtllm_fp4_block_scale_moe` (deferred MoE, shapes `(B, 0)`, B in 1..32) | 128 | 0 |
| `trtllm_batch_decode_mla` (MLA, out of scope here) | 38 | 38 |

The refactor changes how warmup reaches the deferred path, from a direct MoE dispatcher call to the normal full-model dummy pass. The target `(B, 0)` profiles and expected capture-time cache hits are unchanged.

All pre-commit hooks pass on the changed files.

---

AI assistance disclosure: this change was split out of #54507 and adapted to current main with AI assistance from Kimi Code and OpenAI Codex. The human submitter remains responsible for reviewing the final diff and validating the reported results.

### CODE REVIEW

**Comment by mingg26** [comment]
@zyongye  take a look please!

**Review by wzhao18 — COMMENTED** [review]
Thanks for the PR!

Can we simplify this PR by having a helper which simply extracts the `defer_moe_finalize_max_num_tokens` from the MoE layers, and append them to `_flashinfer_autotune_token_counts` and set `tuning_buckets` based on the `num_tokens` for tuning?

This way we keep the same structure as before and merely adds more forward passes with different batch sizes, and avoid MoE-specific calls like:
```
routed_experts.forward_monolithic(
                x=hidden_states,
                router_logits=router_logits,
            )
```

**Comment by mingg26** [comment]
@wzhao18 

Thanks for the suggestion. That would simplify the implementation. 

My concern is the cold-start cost. This runs the entire 93-layer TP8 model under the autotune context, rather than invoking only the affected MoE dispatcher. In an earlier related Kimi-K3 experiment, adding a second full-model autotune pass increased engine initialization from 999.25s to 1281.57s, about 282s or 4.7 minutes.

This is essentially a trade-off between a simpler, more maintainable implementation and additional startup time. I would appreciate your guidance on which side you think we should prioritize.

**Comment by wzhao18** [comment]
@mingg26 We don't want to incur extra engine startup time, I wonder why extra forward passes increases the startup time significantly than MoE-only? The extra tuning part should be the same right? the forward pass itself without tuning should be negligible time.

**Comment by mingg26** [comment]
@wzhao18 You are right. I was conflating the cost of the additional full-model forward with the additional tuning performed in the earlier experiment.  I have updated the PR following your suggestion.

**Comment by wzhao18** [comment]
Looks good to me! Will let @zyongye do a pass and merge this. Thanks!

**Review by zyongye — APPROVED** [review]
